enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,356 @@
+  specific configuration down to components; for example:
+
+  `/etc/microshift/components.d/010-storage-topolvm/config.yaml`
+
+   such yaml would parsed and passed down on the templates of the following
+   point as `{{ .config.value }}` or `{{ .config.sub.value }}`.
+
+*  Kustomize yamls would be pre-processed with go templates to allow insertion of
+  variables as we do today for core assets, i.e.: `{{ .KubeconfigDir }}`,
+  `{{ .BaseDomain }}`,
+  [etc...](https://github.com/openshift/microshift/blob/eb0488fd167333256c69f4eeedc5463c4035a532/pkg/components/render.go#L24).
+  Image references will be injected in the `images` section
+  of the kustomize format not being embedded in `microshift/pkg/releases.go`
+  or passed down as template variables anymore.
+
+* The functionality used in the topo-lvm component to embed the lvmd config file
+  in a config map could be performed with a config map generator, which would contain
+  a default config file. The default config file could be overriden via a custom component
+  variable.
+
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovnk`, `microshift-storage-topolvm`.
+
+* The `microshift-networking-ovnk` package would include the necessary
+  `/usr/lib/microshift/components.d/05-networking-ovnk/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/10-storage-topolvm/` yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as kustomize yamls instead of embedded assets.
+
+* Incompatible component rpms should be impossible to install together.
+  i.e. `microshift-networking-bridge` plus `microshift-networking-ovnk`,
+  this can be handled with Conflict spec clauses. (This is just an example,
+  since microshift-networking-bridge would not be necessary as it's the
+  default for cri-o)
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the `User Stories` section.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+There is an open question about how to handle migration of CRD based
+content between versions. This problem is not handled or explored in MicroShift
+yet. Could this be handled by one-shot pods, the one-shot pods could be defined to
+be triggered only if an specific type of object/version existed, another
+alternative to this would be to use upgrade-binaries optional in components
+which are ran when necessary (similar to how CNIs work).
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+Specific component configuration could be handled as yaml files in
```

Comment:
I don't think we need to specify exactly how a plug-in is configured, other than giving recommendations.

For example, authors could put a ConfigMap resource into a `/etc/microshift/component.d/configmap.yaml` referenced by the kustomization, or they could put an `/etc/microshift-component.d/lvmd.conf` in another format there that gets auto-converted into a ConfigMap through a ConfigMapGenerator.

Reply:
Yeah, that could be an option. But then how do we handle defaults vs parameters set in a configmap like this? you would have the default embedded in the component and the configmap set by the user... Or at boot they would overwrite each other in sequence. I think we still need methods to do this in some cases and I prefer to define how would it work (i.e. ovn specific configuration, it doesn't come via configmaps)